api/envoy/extensions/transport_sockets/tls/v3/common.proto

```diff
@@ -151,7 +151,11 @@ message TlsCertificate {
   // TLS private key is not password encrypted.
   config.core.v3.DataSource password = 3 [(udpa.annotations.sensitive) = true];
 
-  // [#not-implemented-hide:]
+  // The OCSP response to be stapled with this certificate during the handshake.
+  // The response must be DER-encoded if provided as a ``filename`` or
+  // ``inline_bytes``, and PEM-encoded if provided through ``inline_string``.
+  // The response must be unexpired as of configuration time and pertain to
```

Comment:
`The response must be unexpired as of configuration time`. What if Envoy crashes and when it restarts it has expired? Does that prevent restarting? What if there is an unrelated LDS change, but now the staple has expired? Does that cause the LDS update to get rejected? 

Having a config that is valid at one point in time and invalid 30 minutes later seems problematic. I think the behavior to reject config based on this should be optional.

Reply:
Yes this is a bit problematic. Since envoy defaults to not check cert expiration at config time, it seems most consistent to permit expired certs at config time here too. This means any config that is valid at runtime would be valid on a future restart. This also clarifies the `OcspStaplePolicy` since it only now applies to new connections and doesn't imply any config-time requirement:
- `SKIP_STAPLING_IF_REQUIRED` = Optional / Omit but still use the cert if response is expired
- `STAPLING_REQUIRED` = Required / Do not use the cert if response is expired
- `REJECT_CONNECTION_ON_EXPIRED` = Optional / Do not use the cert if response is expired

Reply:
While we're at it, Stephan pointed out the policy names could be clearer, and are also confusingly ordered since the strictest policy is in the middle. How about:
- `LENIENT_STAPLING` = Optional / If expired, use the cert but omit the staple
- `STRICT_STAPLING` = Optional / If expired, do not use the cert for new connections
- `MUST_STAPLE` = Required / If expired, do not use the cert for new connections

Reply:
@ggreenway Should we move forward with the above policies (and no config expiration checks) or would you like to discuss alternatives?

Reply:
I agree with allowing the configuration to load, but rejecting new connections, if we need to staple but cannot.

I don't think I understand the different use case for MUST vs STRICT. Would someone ever specify STRICT but not set an ocsp staple value (expired or otherwise)?

Reply:
My idea was that as an operator introduces OSCP they'd start with LENIENT_STAPLING as they gain confidence in their ability to reliably provide an up-to-date OCSP response in the config. If there's a problem, Envoy continues without stapling, but no outage occurs.

When ready, they switch to STRICT_STAPLING which allows the possibility of removing the OCSP response from configs in an emergency to return to the non-OSCP state, but otherwise requires the response to be up-to-date.

Finally, if they enable the must-staple extension in their SSL cert, they can choose to set MUST_STAPLE in Envoy. In this mode, if the OCSP response is missing or expired, connections will be rejected instead of returning a stale or no OCSP response. The client should reject this OCSP response anyway because the certificate demands its presence. I consider this to be a security-above-all-else setting where you'd rather reject a connection than serve a response without OCSP that the client might choose to ignore (e.g. user bypasses warning).

Reply:
I've merged the change to the enum names and their updated descriptions. Happy to further consider altering these.